modules/guides/src/Access/GuidesAccessCheck.php

```diff
@@ -0,0 +1,110 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+ *  USA.
+ */
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Access\AccessResultInterface;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Routing\Access\AccessInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\guides\GuidesInterface;
+
+/**
+ * Class GuidesAccessCheck.
+ */
+class GuidesAccessCheck implements AccessInterface {
+
+  /**
+   * The guides storage.
+   *
+   * @var \Drupal\guides\GuidesInterface
+   */
+  protected $guidesStorage;
+
+  /**
+   * The cached backend.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cacheBackend;
+
+  /**
+   * GuidesAccessCheck constructor.
+   *
+   * @param \Drupal\guides\GuidesInterface $guides_storage
+   *   The guides storage.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   The cached backend.
+   */
+  public function __construct(GuidesInterface $guides_storage, CacheBackendInterface $cache_backend) {
+    $this->guidesStorage = $guides_storage;
+    $this->cacheBackend = $cache_backend;
+  }
+
+  /**
+   * Grant access to the guides based on the existence of the files.
+   *
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match interface.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The currently logged in account.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   The access result.
+   */
+  public function access(RouteMatchInterface $route_match, AccountInterface $account): AccessResultInterface {
+    if ($account->hasPermission('access guides')) {
+      $file = [];
+
+      $cid = 'guides_files:' . $this->guidesStorage->getDirectory();
+      $data_cached = $this->cacheBackend->get($cid);
+
+      // If files exist in cache then get them from there.
+      if ($data_cached) {
+        $guides = $data_cached->data;
+      }
+      else {
+        $guides = $this->guidesStorage->getFiles();
+      }
+
+      foreach ($guides as $guide) {
+        // Put together the path of the current file.
+        $guides_dir = dirname($guide);
+        $file_path = $guides_dir . '/' . $route_match->getParameter('filename') . '.md';
+        if (file_exists($file_path)) {
+          $file[] = $guide;
+          break;
+        }
+      }
+
+      if ($file) {
+        return AccessResult::allowed();
+      }
+      return AccessResult::forbiddenIf(!$file);
```

Comment:
I'm not really seeing why is it beneficial to return a 403 here in case of a 404. IOW: I think we should return "page not found" when the file is not present, not "access denied".

If I'm wrong, then we should provide a reason for this forbidden result, at least (via the second parameter for `forbiddenIf()`).